app/addons/documents/doc-editor/components.react.jsx

```diff
@@ -0,0 +1,540 @@
+define([
+  'api',
+  'app',
+  'react',
+  'addons/documents/doc-editor/actions',
+  'addons/documents/doc-editor/stores',
+  'addons/fauxton/components.react',
+  'addons/components/react-components.react',
+  'helpers'
+], function (FauxtonAPI, app, React, Actions, Stores, FauxtonComponents, GeneralComponents, Helpers) {
+
+  var store = Stores.docEditorStore;
+
+
+  var DocEditorController = React.createClass({
+
+    getInitialState: function () {
+      return this.getStoreState();
+    },
+
+    getStoreState: function () {
+      return {
+        isLoading: store.isLoading(),
+        doc: store.getDoc(),
+        cloneDocModalVisible: store.isCloneDocModalVisible(),
+        uploadModalVisible: store.isUploadModalVisible(),
+        deleteDocModalVisible: store.isDeleteDocModalVisible()
+      };
+    },
+
+    getDefaultProps: function () {
+      return {
+        database: {},
+        previousPage: '',
+        isNewDoc: false
+      };
+    },
+
+    getCodeEditor: function () {
+      if (this.state.isLoading) {
+        return (<GeneralComponents.LoadLines />);
+      }
+
+      var code = JSON.stringify(this.state.doc.attributes, null, '  ');
+      var editorCommands = [{
+        name: 'save',
+        bindKey: { win: 'Ctrl-S', mac: 'Ctrl-S' },
+        exec: this.saveDoc
+      }];
+
+      return (
+        <GeneralComponents.CodeEditor
+          id="docEditor"
+          ref="docEditor"
+          defaultCode={code}
+          mode="json"
+          autoFocus={true}
+          editorCommands={editorCommands}
+          notifyUnsavedChanges={true}
+          stringEditModalEnabled={true}
+          change={this.onChangeDoc} />
+      );
+    },
+
+    onChangeDoc: function (doc) {
+      // super ugly, but necessary. This tells the user they can't delete the _id or _rev fields as they type and actually
+      // prevents it from being removed in the editable doc
```

Comment:
is there a reason why we keep this?

sometimes i want to create a new document by deleting the id and rev